middleware/index.js

```diff
@@ -0,0 +1,227 @@
+      this.name = props.name;
+      done();
+    }.bind(this));
+
+  },
+
+  askForPhase: function() {
+    var done = this.async();
+
+    var displayName = chalk.yellow(this.name);
+
+    var prompts = [
+      {
+        name: 'phase',
+        message: 'Select the phase for ' + displayName + ':',
+        type: 'list',
+        default: 'routes',
+        choices: listChoices(this.phases.concat(['__other__']))
+      },
+      {
+        name: 'customPhase',
+        message: 'Enter the phase name:',
+        validate: validateName,
+        when: function(answers) {
+          return answers.phase === '__other__';
+        }
+      },
+      {
+        name: 'nextPhase',
+        message: 'Select the phase right after the custom phase:',
+        type: 'list',
+        default: 'routes',
+        choices: listChoices(this.phases.concat(['__last__'])),
+        when: function(answers) {
+          return answers.phase === '__other__';
+        }
+      }
+    ];
+
+    this.prompt(prompts, function(props) {
+      this.phase = props.customPhase || props.phase;
+      this.customPhase = props.customPhase;
+      this.nextPhase = props.nextPhase;
+      done();
+    }.bind(this));
+  },
+
+  promptForPaths: function() {
+    var displayName = chalk.yellow(this.name);
+    this.log('Specify paths for ' + displayName + ':');
+  },
+
+  askForPaths: function() {
+    var done = this.async();
+    this.log('Enter an empty path name when done.');
+    var prompts = [
+      {
+        name: 'path',
+        message: 'Path uri:',
+        validate: function(input) {
+          if (input) {
+            if (input.indexOf('/') !== 0) {
+              return 'Path must start with /';
+            } else {
+              return true;
+            }
+          } else {
+            return true;
+          }
+        }
+      }
+    ];
+    this.prompt(prompts, function(answers) {
+      if (answers.path == null || answers.path === '') {
+        return done();
+      }
+
+      this.paths = this.paths || [];
+      this.paths.push(answers.path);
+      this.log('Let\'s add another path.');
+      this.askForPaths();
+    }.bind(this));
+  },
+
+  askForParams: function() {
+    var done = this.async();
+    var prompts = [
+      {
+        name: 'params',
+        message: 'Configuration parameters in JSON format:',
+        default: '{}',
+        validate: function(input) {
+          if (input) {
+            try {
+              JSON.parse(input);
+            } catch (err) {
+              return err;
+            }
+            return true;
+          } else {
+            return true;
+          }
+        }
+      }
+    ];
+    this.prompt(prompts, function(answers) {
+      this.params = JSON.parse(answers.params);
+      done();
+    }.bind(this));
+  },
+
+  middleware: function() {
+    var done = this.async();
+    var config = {
+      name: this.name,
+      phase: this.phase,
+      paths: this.paths,
+      params: this.params || {},
```

Comment:
Also note that `params` is an array of arguments to pass to the factory function. The case where `params` is an object is a special syntax sugar. It may be worth to use `[]` as default instead of `{}`.